dkg-runtime-primitives/src/proposal.rs

```diff
@@ -30,52 +27,98 @@ pub use webb_proposals::{
+	/// The number of voters
+	pub voter_count: u32,
+	/// The refresh nonce for the rotation
 	pub nonce: ProposalNonce,
+	/// The public key of the governor
 	pub pub_key: Vec<u8>,
 }
 
+impl RefreshProposal {
+	/// Length of the proposal in bytes.
+	pub const LENGTH: usize = 32 + 8 + 4 + ProposalNonce::LENGTH + 64;
+
+	/// Build a `RefreshProposal` from raw bytes
+	pub fn from(bytes: &[u8]) -> Result<Self, &'static str> {
+		Decode::decode(&mut &bytes[..]).map_err(|_| "Failed to decode RefreshProposal")
+	}
+}
+
+impl Default for RefreshProposal {
+	fn default() -> Self {
+		Self {
+			voter_merkle_root: [0u8; 32],
+			session_length: 0,
+			voter_count: 0,
+			nonce: ProposalNonce(0u32),
+			pub_key: vec![],
+		}
+	}
+}
+
+impl MaxEncodedLen for RefreshProposal {
+	fn max_encoded_len() -> usize {
+		Self::LENGTH
+	}
+}
+
 impl Decode for RefreshProposal {
 	fn decode<I: codec::Input>(input: &mut I) -> Result<Self, codec::Error> {
 		const NONCE_LEN: usize = ProposalNonce::LENGTH;
-		let mut data = [0u8; NONCE_LEN + 64];
+		let mut data = [0u8; 32 + 8 + 4 + NONCE_LEN + 64];
 		input.read(&mut data).map_err(|_| {
 			codec::Error::from("input bytes are less than the expected size (68 bytes)")
 		})?;
 		// _NOTE_: rustc won't generate bounds check for the following slice
 		// since we know the length of the slice is at least 68 bytes already.
+		let mut voter_merkle_root_bytes = [0u8; 32];
+		let mut session_length_bytes = [0u8; 8];
+		let mut voter_count_bytes = [0u8; 4];
 		let mut nonce_bytes = [0u8; NONCE_LEN];
 		let mut pub_key_bytes = [0u8; 64];
-		nonce_bytes.copy_from_slice(&data[0..NONCE_LEN]);
-		pub_key_bytes.copy_from_slice(&data[NONCE_LEN..]);
+		voter_merkle_root_bytes.copy_from_slice(&data[0..32]);
+		session_length_bytes.copy_from_slice(&data[32..40]);
+		voter_count_bytes.copy_from_slice(&data[40..44]);
+		nonce_bytes.copy_from_slice(&data[44..(44 + NONCE_LEN)]);
+		pub_key_bytes.copy_from_slice(&data[(44 + NONCE_LEN)..]);
+		let voter_merkle_root = voter_merkle_root_bytes;
+		let session_length = u64::from_be_bytes(session_length_bytes);
+		let voter_count = u32::from_be_bytes(voter_count_bytes);
 		let nonce = ProposalNonce::from(nonce_bytes);
 		let pub_key = pub_key_bytes.to_vec();
-		Ok(Self { nonce, pub_key })
+		Ok(Self { voter_merkle_root, session_length, voter_count, nonce, pub_key })
 	}
 }
 
 impl Encode for RefreshProposal {
 	fn encode(&self) -> Vec<u8> {
 		const NONCE_LEN: usize = ProposalNonce::LENGTH;
-		let mut ret = [0u8; NONCE_LEN + 64];
+		let mut ret = [0u8; 32 + 8 + 4 + NONCE_LEN + 64];
+		let voter_merkle_root = self.voter_merkle_root;
+		let session_length = self.session_length.to_be_bytes();
+		let voter_count = self.voter_count.to_be_bytes();
 		let nonce = self.nonce.to_bytes();
-		ret[0..NONCE_LEN].copy_from_slice(&nonce);
-		ret[NONCE_LEN..(NONCE_LEN + 64)].copy_from_slice(&self.pub_key);
+		let pub_key = self.pub_key.as_slice();
+		ret[0..32].copy_from_slice(&voter_merkle_root);
```

Comment:
For lines 107-111, we are copying bytes to an array. Finally, on line 112, we are heap allocating those bytes behind a `Vec`. This would perform a memcpy behind the scenes, which is not necessary in this case. Instead of declaring an array, you can declare a `vec![0u8; Self::LENGTH]` which already has a zeroed-out vec. Thus, we copy the bytes directly to where they need to in memory without a conversion. When finished, you just return that vec.

Reply:
Related: https://doc.rust-lang.org/beta/src/alloc/vec/mod.rs.html#3119

Reply:
Thank you for the feedback I will update.